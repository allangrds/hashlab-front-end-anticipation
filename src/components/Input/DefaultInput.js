import React from 'react'
import PropTypes from 'prop-types'

const DefaultInput = ({
  className,
  id,
  name,
  onChange,
  type,
}) => {
  function handleOnChange (event) {
    const { value } = event.target
    const { name: inputName } = event.target

    onChange({
      name: inputName,
      value,
    })
  }

  return (
    <input
      id={id}
      className={className}
      onChange={handleOnChange}
      name={name}
      type={type}
    />
  )
}

DefaultInput.propTypes = {
  className: PropTypes.node.isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  type: PropTypes.string.isRequired,
}

export default DefaultInput
