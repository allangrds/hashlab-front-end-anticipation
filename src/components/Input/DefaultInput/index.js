import React from 'react'
import PropTypes from 'prop-types'

const DefaultInput = ({
  className,
  name,
  onChange,
  type,
  value,
  ...rest
}) => {
  function handleOnChange (event) {
    const { value: inputValue } = event.target
    const { name: inputName } = event.target

    onChange({
      name: inputName,
      value: inputValue,
    })
  }

  return (
    <input
      id={name}
      className={className}
      onChange={handleOnChange}
      name={name}
      type={type}
      {...rest}
    />
  )
}

DefaultInput.propTypes = {
  className: PropTypes.node.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  type: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
}

DefaultInput.defaultProps = {
  value: null,
}

export default DefaultInput
