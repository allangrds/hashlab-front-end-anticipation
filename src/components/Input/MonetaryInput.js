import React from 'react'
import PropTypes from 'prop-types'
import CurrencyInput from 'react-currency-input'

const MonetaryInput = ({
  className,
  id,
  name,
  onChange,
  value,
}) => {
  function handleOnChange (event, maskedValue, floatValue) {
    onChange({
      name,
      value: floatValue,
    })
  }

  return (
    <CurrencyInput
      decimalSeparator=","
      thousandSeparator="."
      prefix="R$"
      id={id}
      className={className}
      onChangeEvent={handleOnChange}
      value={value}
    />
  )
}

MonetaryInput.propTypes = {
  className: PropTypes.node.isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
}

MonetaryInput.defaultProps = {
  value: 0,
}

export default MonetaryInput
