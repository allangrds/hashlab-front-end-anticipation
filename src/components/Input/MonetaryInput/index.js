import React from 'react'
import PropTypes from 'prop-types'
import CurrencyInput from 'react-currency-input'

const MonetaryInput = ({
  className,
  name,
  onChange,
  value,
}) => {
  function handleChange (event, maskedValue, floatvValue) {
    const newValue = floatvValue
      ? parseInt(floatvValue.toString().replace(/[^\w\s]/gi, ''), 10)
      : floatvValue

    const payload = {
      name,
      value: newValue,
    }

    onChange(payload)
  }

  return (
    <CurrencyInput
      onChangeEvent={handleChange}
      decimalSeparator=","
      thousandSeparator="."
      prefix="R$"
      id={name}
      className={className}
      value={value}
    />
  )
}

MonetaryInput.propTypes = {
  className: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
}

MonetaryInput.defaultProps = {
  className: null,
  value: null,
}

export default MonetaryInput
