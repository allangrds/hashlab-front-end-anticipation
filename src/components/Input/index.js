import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import MonetaryInput from './MonetaryInput'
import DefaultInput from './DefaultInput'
import styles from './styles.module.css'

const Input = ({
  className,
  errorMessage,
  id,
  label,
  name,
  onChange,
  required,
  tip,
  type,
  value,
}) => {
  const bottomText = errorMessage || tip
  const bottomClassName = errorMessage
    ? styles.error
    : styles.tip
  const inputClassName = errorMessage
    ? classNames(styles.input, styles['input-error'])
    : styles.input
  const SelectedInput = type === 'money'
    ? MonetaryInput
    : DefaultInput
  const fieldsetClassName = className
    ? classNames(styles.fieldset, className)
    : styles.fieldset

  function handleOnChange (obj) {
    onChange(obj)
  }

  return (
    <fieldset className={fieldsetClassName}>
      <label
        className={styles.label}
        htmlFor={id}
      >
        { label } { required && '*' }
      </label>
      <SelectedInput
        className={inputClassName}
        id={id}
        name={name}
        onChange={handleOnChange}
        type={type}
        value={value}
      />
      <p className={bottomClassName}>
        { bottomText }
      </p>
    </fieldset>
  )
}

Input.propTypes = {
  className: PropTypes.node.isRequired,
  errorMessage: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  required: PropTypes.bool,
  tip: PropTypes.string,
  type: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
}

Input.defaultProps = {
  errorMessage: null,
  id: null,
  required: false,
  tip: null,
  value: null,
}

export default Input
