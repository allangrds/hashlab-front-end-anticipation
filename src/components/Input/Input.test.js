import React from 'react'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import Input from '.'

test('Input with type "text" should render and be updated correctly', () => {
  const callback = jest.fn()
  const props = {
    id: 'idLegal',
    label: 'Esse é o label',
    name: 'input',
    onChange: callback,
    required: true,
    tip: 'Máximo de 12 parcelas',
    type: 'text',
  }
  const value = 'a'

  const {
    container,
    getByDisplayValue,
    getByLabelText,
    getByText,
  } = render(
    <Input {...props} />,
  )

  expect(callback).not.toHaveBeenCalled()

  getByLabelText(`${props.label} *`)
  getByText(props.tip)

  const input = getByDisplayValue('')

  userEvent.type(input, value)

  expect(callback).toHaveBeenCalledTimes(1)

  expect(container).toMatchSnapshot()
})

test('Input with type "money" should render and be updated correctly', () => {
  const callback = jest.fn()
  const props = {
    id: 'oId',
    label: 'Esse é o label',
    name: 'input',
    onChange: callback,
    required: true,
    tip: 'Máximo de 12 parcelas',
    type: 'money',
  }
  const initialValue = 'R$0,00'
  const newRawValue = '123'
  const newFormatedValue = 'R$1,23'

  const {
    container,
    getByDisplayValue,
    getByLabelText,
    getByText,
  } = render(
    <Input {...props} />,
  )

  expect(callback).not.toHaveBeenCalled()

  getByLabelText(`${props.label} *`)
  getByText(props.tip)

  const input = getByDisplayValue(initialValue)

  userEvent.type(input, newRawValue)

  expect(callback).toHaveBeenCalledTimes(3)

  getByDisplayValue(newFormatedValue)

  expect(container).toMatchSnapshot()
})

test('Input should render with error message', () => {
  const props = {
    errorMessage: 'Mensagem de erro',
    id: 'oId',
    label: 'Esse é o label',
    name: 'input',
    onChange: () => null,
    required: true,
    type: 'text',
  }

  const {
    container,
    getByText,
  } = render(
    <Input {...props} />,
  )

  getByText(props.errorMessage)

  expect(container).toMatchSnapshot()
})
