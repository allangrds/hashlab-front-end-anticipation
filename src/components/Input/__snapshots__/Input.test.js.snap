// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render with error message 1`] = `
<div>
  <fieldset
    class="fieldset"
  >
    <label
      class="label"
      for="oId"
    >
      Esse é o label
       
      *
    </label>
    <input
      class="input input-error"
      id="oId"
      name="input"
      type="text"
    />
    <p
      class="error"
    >
      Mensagem de erro
    </p>
  </fieldset>
</div>
`;

exports[`Input with type "money" should render and be updated correctly 1`] = `
<div>
  <fieldset
    class="fieldset"
  >
    <label
      class="label"
      for="oId"
    >
      Esse é o label
       
      *
    </label>
    <input
      class="input"
      id="oId"
      type="text"
      value="R$1,23"
    />
    <p
      class="tip"
    >
      Máximo de 12 parcelas
    </p>
  </fieldset>
</div>
`;

exports[`Input with type "text" should render and be updated correctly 1`] = `
<div>
  <fieldset
    class="fieldset"
  >
    <label
      class="label"
      for="idLegal"
    >
      Esse é o label
       
      *
    </label>
    <input
      class="input"
      id="idLegal"
      name="input"
      type="text"
    />
    <p
      class="tip"
    >
      Máximo de 12 parcelas
    </p>
  </fieldset>
</div>
`;
